javalib/src/main/scala/java/lang/process/GenericProcess.scala

```diff
@@ -1,7 +1,163 @@
+      val cmd = builder.command()
+      val cmdAsArray = cmd.toArray(new Array[String](cmd.size()))
+      val command =
+        if (cmd.isEmpty()) Optional.empty[String]()
+        else Optional.of(cmdAsArray.head)
+      val args =
+        if (cmd.isEmpty()) Optional.empty[Array[String]]()
+        else Optional.of(cmdAsArray.tail)
+      new Info(pid = pid, cmd = command, args = args)
+    }
+  }
+  class Info(
+      val pid: scala.Long,
+      cmd: Optional[String],
+      args: Optional[Array[String]]
+  ) extends ProcessHandle.Info {
+    override def command(): Optional[String] = cmd
+    override def arguments(): Optional[Array[String]] = args
+    override def commandLine(): Optional[String] =
+      // For comprehension variant does not compile on Scala 2.12
+      command().flatMap[String] { cmd =>
+        arguments().map[String] { args =>
+          s"$cmd ${args.mkString(" ")}"
+        }
+      }
+
+    override def user(): Optional[String] =
+      Optional.ofNullable(System.getProperty("user.name"))
+
+    // Instant not implemented
+    override def startInstant(): Optional[Instant] = Optional.empty()
+    override def totalCpuDuration(): Optional[Duration] = Optional.empty()
+
+    override def toString(): String = {
+      val args = this
+        .arguments()
+        .orElseGet(() => Array[String]())
+        .asInstanceOf[Array[AnyRef]]
+      s"[user: ${user()}, cmd: ${{ command() }}, args: ${Arrays.toString(args)}"
+    }
+  }
+
+  object ProcessWatcher {
+    // Identity map, no valid concurrent structure
+    val watchedProcesses =
+      new ConcurrentHashMap[CompletableFuture[
+        ProcessHandle
+      ], GenericProcess.Handle]()
+
+    private val lock = new ju.concurrent.locks.ReentrantLock()
+    private val hasProcessesToWatch = lock.newCondition()
+
+    def watchForTermination(
+        completion: CompletableFuture[ProcessHandle],
+        handle: GenericProcess.Handle
+    ): Unit = {
+      watchedProcesses.put(completion, handle)
+      assert(
+        watcherThread.isAlive(),
+        "Process termination watch thread is terminated"
+      )
+      // If we cannot lock it means that watcher thread is already executing, no need to wake it up
+      if (lock.tryLock()) {
+        try hasProcessesToWatch.signal()
+        finally lock.unlock()
+      }
+    }
+
+    lazy val watcherThread: Thread = Thread
+      .ofPlatform()
+      .daemon(true)
+      .group(ThreadGroup.System)
+      .name("ScalaNative-ProcessTerminationWatcher")
+      .startInternal { () =>
+        lock.lock()
+        try {
+          while (true) try {
+            while (watchedProcesses.isEmpty()) {
+              hasProcessesToWatch.await()
+            }
+            watchedProcesses.forEach { (completion, handle) =>
+              if (completion.isCancelled())
+                watchedProcesses.remove(completion)
+              else if (!handle.isAlive() ||
```

Comment:
I think the `handle.isAlive()` call can be removed if the Windows
implementaton of `isAlive()` and `waitFor()` permit.

On `UnixProcessGen2` `isAlive()` will call into `waitPidImpl()` with
a `options = WNOHANG" if it does not know that the child
process has already exited.  If the `waitfor(WNOHANG)` indicates
that the child has not exited, then the right half of the "||" will
get executed causing a "waitfor(timeout)".  Might as well do
only one trip all the way to the OS rather than two.

Reply:
Good catch, thanks